akka-contrib/src/test/scala/akka/contrib/circuitbreaker/CircuitBreakerProxySpec.scala

```diff
@@ -0,0 +1,536 @@
+      sender.send(circuitBreaker, Passthrough("This should be received"))
+      receiver.expectMsg("This should be received")
+
+      And("The circuit is still closed for ordinary messages")
+      sender.send(circuitBreaker, "This should NOT be received")
+      receiver.expectNoMsg()
+    }
+  }
+
+  "Ask Extension" should {
+    implicit val patienceConfig = PatienceConfig(timeout = Span(1, Second), interval = Span(100, Millis))
+    import Implicits.askWithCircuitBreaker
+
+    import scala.concurrent.ExecutionContext.Implicits.global
+    implicit val timeout: Timeout = 2.seconds
+
+    "work as a ASK pattern if circuit is closed" in new CircuitBreakerScenario {
+      Given("A circuit breaker actor proxying a test probe")
+      val circuitBreaker = defaultCircuitBreaker
+
+      When("Doing a askWithCircuitBreaker request")
+      val responseFuture = circuitBreaker.askWithCircuitBreaker("request")
+
+      Then("The message is sent to the target actor")
+      receiver.expectMsg("request")
+
+      When("Then target actor replies")
+      receiver.reply("response")
+
+      Then("The response is available as result of the future returned by the askWithCircuitBreaker method")
+      whenReady(responseFuture) { response ⇒
+        response should be("response")
+      }
+    }
+
+    "transform the response into a failure with CircuitOpenException cause if circuit is open" in new CircuitBreakerScenario {
+      Given("A circuit breaker actor proxying a test probe")
+      val circuitBreaker = defaultCircuitBreaker
+
+      When("The circuit breaker proxy enters OPEN state")
+      receiverRespondsWithFailureToRequest("request1")
+      receiverRespondsWithFailureToRequest("request2")
+
+      circuitBreakerReceivesSelfNotificationMessage()
+
+      And("Doing a askWithCircuitBreaker request")
+      val responseFuture = circuitBreaker.askWithCircuitBreaker("request")
+
+      Then("The message is NOT sent to the target actor")
+      receiver.expectNoMsg()
+
+      And("The response is converted into a failure")
+      whenReady(responseFuture.failed) { failure ⇒
+        failure shouldBe a[OpenCircuitException]
+      }
+    }
+  }
+
+  "Future Extension" should {
+    implicit val patienceConfig = PatienceConfig(timeout = Span(1, Second), interval = Span(100, Millis))
+    import Implicits.futureExtensions
+    import akka.pattern.ask
+
+    import scala.concurrent.ExecutionContext.Implicits.global
+    implicit val timeout: Timeout = 2.seconds
+
+    "work as a ASK pattern if circuit is closed" in new CircuitBreakerScenario {
+      Given("A circuit breaker actor proxying a test probe")
+      val circuitBreaker = defaultCircuitBreaker
+
+      When("Doing a askWithCircuitBreaker request")
+      val responseFuture = (circuitBreaker ? "request").failForOpenCircuit
+
+      Then("The message is sent to the target actor")
+      receiver.expectMsg("request")
+
+      When("Then target actor replies")
```

Comment:
Isn't the whole idea with given-when-then that that is the entire list of states, here it repeats.